homeassistant/components/modbus/base_platform.py

```diff
@@ -106,6 +108,7 @@ def get_optional_numeric_config(config_name: str) -> int | float | None:
         self._min_value = get_optional_numeric_config(CONF_MIN_VALUE)
         self._max_value = get_optional_numeric_config(CONF_MAX_VALUE)
         self._zero_suppress = get_optional_numeric_config(CONF_ZERO_SUPPRESS)
+        self._nan_value_hex = get_optional_numeric_config(CONF_NAN_VALUE)
```

Comment:
NAN is only defined for sensors, but this is a base class for more than sensors e.g. climate.

please Either move this to the right class or correct the configuration.

Reply:
> I still have an issue, it's basically how to input a hex value via the yaml configuration.
> 
> ```
> modbus:
>   - type: tcp
>     host: IP_ADDRESS_1
>     port: 2020
>     name: "hub1"
>     sensors:
>       - name: sensor1
>         data_type: int32
>         nan_value: "0x80000000"
> ```
> 
> It can't be automatically read in and converted by python, as it would bypass the structs/endianess/sensor dtype and convert it into a wrong integer. Basically we have to read it in as string and later unpack the hex string like
> 
> `struct.unpack(self._structure, bytes.fromhex(str(config[CONF_NAN_VALUE])[2:]))[0] `
> 
> test config would look like this:
> 
> ```
>         (
>             {
>                 CONF_DATA_TYPE: DataType.STRING,
>                 CONF_NAN_VALUE: str("0x80000000"),
>             },
>             [0x8000, 0x0000],
>             False,
>             STATE_UNAVAILABLE,
>         ),
> ```
> 
> but this ends up with an invalid config when running the above test config and haven't figured out why it just doesn't not like strings as input and wants a number
> 
> ```
> hass = <homeassistant.core.HomeAssistant object at 0xffffa1412350>, caplog = <_pytest.logging.LogCaptureFixture object at 0xffffa13a8100>
> register_words = [65535, 65535], check_config_loaded = True, config_addon = {'data_type': <DataType.STRING: 'string'>, 'nan_value': '0xFFFFFFFF'}
> do_config = {'sensors': [{'address': 51, 'name': 'test entity', 'scan_interval': 1}]}
> 
>     @pytest.fixture(name="mock_modbus")
>     async def mock_modbus_fixture(
>         hass, caplog, register_words, check_config_loaded, config_addon, do_config
>     ):
>         """Load integration modbus using mocked pymodbus."""
>         conf = copy.deepcopy(do_config)
>         for key in conf:
>             if config_addon:
>                 conf[key][0].update(config_addon)
>             for entity in conf[key]:
>                 if CONF_SLAVE not in entity:
>                     entity[CONF_SLAVE] = 0
>         caplog.set_level(logging.WARNING)
>         config = {
>             DOMAIN: [
>                 {
>                     CONF_TYPE: TCP,
>                     CONF_HOST: TEST_MODBUS_HOST,
>                     CONF_PORT: TEST_PORT_TCP,
>                     CONF_NAME: TEST_MODBUS_NAME,
>                     **conf,
>                 }
>             ]
>         }
>         mock_pb = mock.MagicMock()
>         with mock.patch(
>             "homeassistant.components.modbus.modbus.ModbusTcpClient",
>             return_value=mock_pb,
>             autospec=True,
>         ):
>             now = dt_util.utcnow()
>             with mock.patch(
>                 "homeassistant.helpers.event.dt_util.utcnow",
>                 return_value=now,
>                 autospec=True,
>             ):
>                 result = await async_setup_component(hass, DOMAIN, config)
> >               assert result or not check_config_loaded
> E               assert (False or not True)
> 
> tests/components/modbus/conftest.py:116: AssertionError
> --------------------------------------------------------------- Captured stderr setup ----------------------------------------------------------------
> WARNING:homeassistant.components.modbus.validators:sensor test entity scan_interval(1) is lower than 5 seconds, which may cause Home Assistant stability issues
> ERROR:homeassistant.config:Invalid config for [modbus]: invalid number 0x80000000 for dictionary value @ data['modbus'][0]['sensors'][0]['nan_value']. Got '0x80000000'. (See ?, line ?). Please check the docs at https://www.home-assistant.io/integrations/modbus
> ERROR:homeassistant.setup:Setup failed for modbus: Invalid config.
> ```

And now I feel stupid. Found it in the `__init__.py`

Reply:
here I don't know, 
`__process_raw_value(self, entry: float | int | str)` is protected in

`class BaseStructPlatform(BasePlatform, RestoreEntity):`

so can't overwrite it in

`class ModbusRegisterSensor(BaseStructPlatform, RestoreSensor, SensorEntity):
`

and it's unclear to me what you mean with otherwise correct the configuration

Reply:
And alo wondering, since I actually followed the min_value/max_value/zero_supress PR logic, which is also just sensor specific.